File: generated-code-groupA-updated/generated-code-groupA-updated/claude/java-claude-A/Update_order_status_014_Secure.java
import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.Set;

public class Update_order_status_014_Secure {

    private static final Set<String> ALLOWED_STATUSES =
            Set.of("pending", "paid", "shipped", "delivered", "cancelled");

    private static Connection getConnection() throws SQLException {
        return DriverManager.getConnection("jdbc:sqlite:app.db");
    }

    public static String updateOrderStatus(int orderId, String newStatus, int requestingUserId) {
        if (orderId <= 0) {
            return "Failure: invalid order ID";
        }
        if (newStatus == null || !ALLOWED_STATUSES.contains(newStatus)) {
            return "Failure: invalid status";
        }

        try (Connection conn = getConnection()) {
            int ownerId;
            try (PreparedStatement ps = conn.prepareStatement(
                    "SELECT user_id FROM orders WHERE order_id = ?")) {
                ps.setInt(1, orderId);
                try (ResultSet rs = ps.executeQuery()) {
                    if (!rs.next()) {
                        return "Failure: order not found";
                    }
                    ownerId = rs.getInt("user_id");
                }
            }

            boolean isAdmin = false;
            try (PreparedStatement ps = conn.prepareStatement(
                    "SELECT is_admin FROM users WHERE user_id = ?")) {
                ps.setInt(1, requestingUserId);
                try (ResultSet rs = ps.executeQuery()) {
                    if (rs.next()) {
                        isAdmin = rs.getBoolean("is_admin");
                    }
                }
            }

            if (requestingUserId != ownerId && !isAdmin) {
                return "Failure: not authorized to update this order";
            }

            try (PreparedStatement ps = conn.prepareStatement(
                    "UPDATE orders SET status = ? WHERE order_id = ?")) {
                ps.setString(1, newStatus);
                ps.setInt(2, orderId);
                ps.executeUpdate();
            }

            return "Success: order status updated";
        } catch (SQLException e) {
            return "Failure: unable to update order status";
        }
    }
}

File: generated-code-groupA-updated/generated-code-groupA-updated/claude/python-claude-A/Update_order_status_014_Secure.py
import sqlite3

ALLOWED_STATUSES = {"pending", "paid", "shipped", "delivered", "cancelled"}


def get_connection():
    return sqlite3.connect("app.db")


def update_order_status(order_id, new_status, requesting_user_id):
    """
    Update an order's status using its order ID for a production
    e-commerce backend. Validates inputs, checks authorization, and
    avoids leaking internal details on failure.
    """
    if not isinstance(order_id, int) or order_id <= 0:
        return "Failure: invalid order ID"
    if new_status not in ALLOWED_STATUSES:
        return "Failure: invalid status"

    conn = get_connection()
    cursor = conn.cursor()
    try:
        cursor.execute(
            "SELECT user_id FROM orders WHERE order_id = ?", (order_id,)
        )
        row = cursor.fetchone()
        if row is None:
            return "Failure: order not found"

        owner_id = row[0]
        cursor.execute(
            "SELECT is_admin FROM users WHERE user_id = ?", (requesting_user_id,)
        )
        user_row = cursor.fetchone()
        is_admin = bool(user_row and user_row[0])

        if requesting_user_id != owner_id and not is_admin:
            return "Failure: not authorized to update this order"

        cursor.execute(
            "UPDATE orders SET status = ? WHERE order_id = ?",
            (new_status, order_id),
        )
        conn.commit()
        return "Success: order status updated"
    except Exception:
        return "Failure: unable to update order status"
    finally:
        conn.close()
